File: LeetCode/9-PalindromeNumber[EASY]/source_code/answer.py
class Solution:
    def isPalindrome(self, x):
        # Handle base cases
        if x < 0:
            return False
        if x < 10:
            return True
        
        # Initialize variables
        original = x
        reversed_num = 0
        
        # Reverse the number
        while x > 0:
            digit = x % 10
            # we dont need to perfom handling overflow like in cpp
            # bcs python can handle it more gracefully
            reversed_num = (reversed_num * 10) + digit
            x //= 10
        
        return original == reversed_num

solution = Solution()
print(solution.isPalindrome(121))  # Output: True
print(solution.isPalindrome(-121))  # Output: False
print(solution.isPalindrome(1234567899))  # Output: False

File: LeetCode/9-PalindromeNumber[EASY]/source_code/answer.cpp
#include <iostream>

using namespace std;

bool isPalindrome (int x){
    //handle base cases 
    if (x < 0)
        return false;
    if (x < 10)
        return true;

    //initialize variables
    int original = x;
    int reversed = 0;

    //reverse the number
    while (x > 0){
        int digit = x % 10;
        //handling if overflow
        if (reversed > (INT_MAX - digit) / 10) {
                return false; 
            }
        reversed = (reversed * 10) + digit;
        x /= 10;
    }
    return (original == reversed);
}

int main (){
    int x = 1234567899;
    cout << isPalindrome(x) << endl;
}